### PR TITLE
[MCC-200787] 2 new options

### PR DESCRIPTION
**Pull request by rcohen-mdsol**
Rack::ContentTypeDefault is a piece of middleware that sets default `CONTENT-TYPE:` header when it isn't present. It checks if a content type is set and does not override the existing one.  Currently, the options allow you to specify the request methods and the content type to be set. So for example, you can decide to set the content-type to 'application/x-www-form-urlencoded' on post requests if one is not provided.

The purpose of this PR is to add two new options for the user without breaking existing functionality. iMedidata will take advantage of these new options in order to set content type if none is provided by the API consumer for requests made via the authenticate route.

@pmavinkurve-mdsol @cwhite-mdsol @sakhtermdsol @fpcyan @njacobs1 

New option 1: The user can decide to set the content type (if one is not provided) only for requests with certain paths. So for example, specifying "authenticate" would only set content type for requests with paths ending in "/authenticate". The user can specify 1, multiple, or all paths. If nothing is specified, the default is that there is no path validation required (ie, that all paths are specified). 

New option 2: The user has the option to have content type set to application/xml or application/json if the path ends with .xml or .json, respectively. If the path ends in neither, the content type is set to the one specified by the user. The default behavior (when the option isn't explicitly exercised) is not to determine content type based on path ending at all, but rather to always set content type to the one specified by the user.

iMedidata will use these two new options to set content type for authentication requests made via the authenticate route with the following:

config.middleware.use Rack::ContentTypeDefault, [:get, :post], 'application/json', 'authenticate', true
 
rspec
Finished in 0.00749 seconds (files took 0.1922 seconds to load)
19 examples, 0 failures

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,3 +1,4 @@
 source "http://rubygems.org"
+gem 'activesupport', '~> 3.2.22'
```
Comment (CGDS) [inline Gemfile]: Why is this needed? Pretty heavy weight dependency to pull in for a relatively small gem, also the rails version lock is going to be challenging.  In any case, gems served by rubygems should be included as a dependency in in the gemspec so bundler can parse dependencies.

File: lib/rack/content_type_default.rb [hunk 2]
```diff
@@ -1,32 +1,62 @@
...
       @app = app
       @methods = [*methods].map{ |item| item.to_s.upcase }
       @content_type = content_type
+      @paths = [*paths]
+      @default_based_on_path = default_based_on_path
     end
 
     def call(env)
       req = Rack::Request.new(env)
+      @app.call(env) and return unless reset_content_type?(req)
```
Comment (CGDS) [inline lib/rack/content_type_default.rb]: So does this end up allowing `@app.call(env)` to be run twice if `reset_content_type` returns false?
```ruby
[1] pry(main)> def low_precendence_and(bool)
[1] pry(main)*   puts("foo") and return unless bool
[1] pry(main)*   "bar"
[1] pry(main)* end
=> :low_precedence_and
[2] pry(main)> low_precedence_and(true)
=> "bar"
[3] pry(main)> low_precedence_and(false)
foo
=> "bar"
```

File: lib/rack/content_type_default.rb [hunk 3]
```diff
@@ -1,32 +1,62 @@
...
-
+      env['CONTENT_TYPE'] = @default_based_on_path ? determine_content_type(req.env['PATH_INFO']) :  @content_type
       @app.call(env)
     end
 
+
+    private
+    # Determine whether or not to reset content type
+    def reset_content_type?(req)
+      return false unless req.content_type.nil? or req.content_type.empty?
```
Comment (CGDS) [inline lib/rack/content_type_default.rb]: these low precedence operators make me nervous.  Also seems like a lot of explicit returns in this method.

File: lib/rack/content_type_default.rb [hunk 4]
```diff
@@ -1,32 +1,62 @@
...
-      if match_method?(req.request_method)
-        env['CONTENT_TYPE'] = @content_type if req.content_type.nil? or req.content_type.empty?
-      end
-
+      env['CONTENT_TYPE'] = @default_based_on_path ? determine_content_type(req.env['PATH_INFO']) :  @content_type
       @app.call(env)
     end
 
+
+    private
```
Comment (CGDS) [inline lib/rack/content_type_default.rb]: Don't need all 4 `private` calls.  All methods defined after `private` are also private

File: lib/rack/content_type_default.rb [hunk 5]
```diff
@@ -1,32 +1,62 @@
...
+    def match_path?(path_info)
+      return true if @paths.first == 'all'
+      return @paths.any? { |path| path_info.ends_with?(path) or path_info.include?("#{path}.") }
+    end
+
+    private
+    #Determine if content type can be set to application/json or application/xml based on the path ending.
+    #If the path does not end in xml or json, content type is set to the given default.
+    def determine_content_type(path_info)
+      return 'application/xml' if path_info.split('.').last == 'xml'
```
Comment (CGDS) [inline lib/rack/content_type_default.rb]: Why do we need to set the ContentType header in these cases ?  Is there a new issue with rails content negotiation?

File: spec/lib/rack/content_type_default_spec.rb [hunk 6]
```diff
@@ -2,7 +2,11 @@
 require 'rack/mock'
 
 describe Rack::ContentTypeDefault do
-  App = lambda { |env| [200, {'Content-Type' => 'text/plain'}, env['PATH_INFO']] }
+  before do
+    App = lambda { |env| [200, {'Content-Type' => 'text/plain'}, env['PATH_INFO']] }
+    @post_req = Rack::MockRequest.env_for("/api/v2/customers", method: :post)
```
Comment (CGDS) [inline spec/lib/rack/content_type_default_spec.rb]: could these be moved to `lets`?

File: spec/lib/rack/content_type_default_spec.rb [hunk 7]
```diff
@@ -15,63 +19,117 @@
...
+      Rack::ContentTypeDefault.new(App, :get, 'application/xml').call(@get_req)
+      @get_req['CONTENT_TYPE'].should == 'application/xml'
     end
   end
-end
 
+  context 'paths' do
+    it 'sets the content type as appropriate if the path in the request matches one specified' do
+      @get_req['PATH_INFO'] = 'api/v2/customers/2155/authenticate.json'
+      Rack::ContentTypeDefault.new(App, :get, 'application/xml', ['/show', '/authenticate.json']).call(@get_req)
```
Comment (CGDS) [inline spec/lib/rack/content_type_default_spec.rb]: So server semantics for a GET with a body are constrained.  Basically, any client is allowed to send one, and if a server receives a get with a body, if any, it must not change the semantics of its response based on that body.  Since a server isn't allowed to change things on that body, the content type specifying the encoding of that body may be discarded.   From some of these tests, it kind of looks like you might need some behavior from [ignore_get_content_type](https://github.com/mdsol/ignore_get_content_type)?

Comment (CGDS) [inline spec/lib/rack/content_type_default_spec.rb]: Sorry, just double checked the spec.  It's a SHOULD not a MUST.

File: lib/rack/content_type_default.rb (hunk 5 above)
Comment (rcohen-mdsol) [inline lib/rack/content_type_default.rb]: @CGDS I previously set content type default to application/json. But when I made requests with paths ending in .xml, content type was still set to application/json.

File: lib/rack/content_type_default.rb (hunk 2 above)
Comment (fpcyan) [inline lib/rack/content_type_default.rb]: What is the difference between this and `return @app.call(env) unless reset_content_type?(req)`?

File: spec/lib/rack/content_type_default_spec.rb [hunk 8]
```diff
@@ -15,63 +17,117 @@
...
+      Rack::ContentTypeDefault.new(App, :get, 'application/xml', 'all').call(get_req)
+      get_req['CONTENT_TYPE'].should == 'application/xml'
+    end
+
+    it 'sets content type as appropriate on all requests if paths is none are explicitly specified' do
+      Rack::ContentTypeDefault.new(App, :post, 'application/xml').call(post_req)
+      post_req['CONTENT_TYPE'].should == 'application/xml'
+    end
+
+    it 'sets content type on no requests if paths specified as empty ' do
```
Comment (fpcyan) [inline spec/lib/rack/content_type_default_spec.rb]: it 'does not set content type if paths specified as empty' ?

**Comment by rcohen-mdsol** [comment]
@CGDS thanks for reviewing :) all comments addressed

**Comment by pmavinkurve-mdsol** [comment]
This is indeed a useful addition. Thanks @rcohen-mdsol! Sorry for the delayed review. 
Thanks @CGDS and @fpcyan for the review! I may have super minor comments that I will make now.

File: README.md [hunk 9]
```diff
@@ -1,8 +1,10 @@
 # Rack::ContentTypeDefault
 
 Rack::ContentTypeDefault is a tiny piece of middleware that sets default `CONTENT-TYPE:` header when it isn't present
-for specific methods. It checks if a content type is set and does not override the existing one.
-The options allow you to specify the request methods, and the content type to be set.
+for specific requests. It checks if a content type is set and does not override the existing one.
+The options allow you to specify the request methods and the content type to be set. The options also allow you to set
```
Comment (pmavinkurve-mdsol) [inline README.md]: minor: How about a line break here. and simply list out all possible options as a numbered or bulleted list. Something like:

The following options may be specified:
* Request methods
* Desired content type
* Paths to which the content type is desired/applicable. `all` being the default when not specified.
* Another content type option that only applies to `application/xml` or `application/json` if the path ends in .xml or .json respectively.

Comment (pmavinkurve-mdsol) [inline README.md]: in fact i would flip 3 & 4. sorry i thought path was the third param. but it's the last.

**Comment by pmavinkurve-mdsol** [comment]
@rcohen-mdsol can you give specific examples where you need the middleware at all for the case of xml and json content type options? I am curious. When we have `.xml` or `.json` doesn't Rails do the right magic anyway?

File: README.md [hunk 10]
```diff
@@ -13,13 +15,23 @@ The code has been copied and adapted from https://gist.github.com/tstachl/626424
...
 ```
 
+```
+require 'rack/content-type-default`
+config.middleware.use Rack::ContentTypeDefault, [:get, :post], 'application/xml', '/authenticate.xml'
+```
+
+```
+require 'rack/content-type-default`
+config.middleware.use Rack::ContentTypeDefault, :post, 'application/json', ['/authenticate.json', '/show'], true
```
Comment (pmavinkurve-mdsol) [inline README.md]: This `true`/false option isn't described in the Readme. Can you please elaborate on it above in the to-be-numbered list?

File: lib/rack/content_type_default.rb [hunk 11]
```diff
@@ -1,32 +1,54 @@
...
-        env['CONTENT_TYPE'] = @content_type if req.content_type.nil? or req.content_type.empty?
+      if set_content_type?(req)
+        env['CONTENT_TYPE'] = @default_based_on_path ? determine_content_type(req.env['PATH_INFO']) :  @content_type
       end
-
       @app.call(env)
     end
 
     private
+    # Determine whether or not to reset content type
```
Comment (pmavinkurve-mdsol) [inline lib/rack/content_type_default.rb]: how about `reset` -> `override` or just `set/override`

File: lib/rack/content_type_default.rb [hunk 12]
```diff
@@ -1,32 +1,54 @@
...
       @methods.include?(method)
     end
+
+    # Match the path on the request with the paths specified. Content type is set only if a match is found.
+    # If @paths was provided with 'all' as the first element (default), content type will be set for all requests.
+    # If @paths was explicitly set to empty, content type will not be set.
+    def match_path?(path_info)
+      return true if @paths.first == 'all'
+      return @paths.any? { |path| /#{path}$/.match(path_info) or /#{path}\./.match(path_info) }
+
```
Comment (pmavinkurve-mdsol) [inline lib/rack/content_type_default.rb]: nit-pick: remove extraneous new line

File: lib/rack/content_type_default.rb [hunk 13]
```diff
@@ -1,32 +1,54 @@
...
+    # If @methods was provided as empty, content type will not be set.
     def match_method?(method)
       @methods.include?(method)
     end
+
+    # Match the path on the request with the paths specified. Content type is set only if a match is found.
+    # If @paths was provided with 'all' as the first element (default), content type will be set for all requests.
+    # If @paths was explicitly set to empty, content type will not be set.
+    def match_path?(path_info)
+      return true if @paths.first == 'all'
```
Comment (pmavinkurve-mdsol) [inline lib/rack/content_type_default.rb]: To make it more ruby-like. If I may suggest an alternate:

`@paths.first == 'all' || @paths.any? { |path| /#{path}$/.match(filter) || /#{path}\./.match(filter) }`

I renamed the `path_info` var to `filter` to fit this into 101 chars but it may still spill beyond 100 with the indentation. 

can leave as-is, or indent the part after the first `||` on a newline.

File: lib/rack/content_type_default.rb [hunk 14]
```diff
@@ -1,32 +1,54 @@
...
+      return @paths.any? { |path| /#{path}$/.match(path_info) or /#{path}\./.match(path_info) }
+
+    end
+
+    #Determine if content type can be set to application/json or application/xml based on the path ending.
+    #If the path does not end in xml or json, content type is set to the given default.
+    def determine_content_type(path_info)
+      return 'application/xml' if path_info.split('.').last == 'xml'
```
Comment (pmavinkurve-mdsol) [inline lib/rack/content_type_default.rb]: again a matter of taste. `return` is typically not used a lot in ruby. an alternate:

```
inferred_content_type = path_info.split('.').last
inferred_content_type == 'json' || 'xml' ? inferred_content_type : @content_type
```

File: spec/lib/rack/content_type_default_spec.rb [hunk 15]
```diff
@@ -3,6 +3,8 @@
 
 describe Rack::ContentTypeDefault do
   App = lambda { |env| [200, {'Content-Type' => 'text/plain'}, env['PATH_INFO']] }
+  let(:post_req) { Rack::MockRequest.env_for("/api/v2/customers", method: :post) }
```
Comment (pmavinkurve-mdsol) [inline spec/lib/rack/content_type_default_spec.rb]: :+1:

File: spec/lib/rack/content_type_default_spec.rb [hunk 16]
```diff
@@ -15,63 +17,117 @@
...
+      Rack::ContentTypeDefault.new(App, :get, 'application/xml').call(get_req)
+      get_req['CONTENT_TYPE'].should == 'application/xml'
     end
   end
-end
 
+  context 'paths' do
+    it 'sets the content type as appropriate if the path in the request matches one specified' do
+      get_req['PATH_INFO'] = 'api/v2/customers/2155/authenticate.json'
+      Rack::ContentTypeDefault.new(App, :get, 'application/xml', ['/show', '/authenticate.json']).call(get_req)
```
Comment (pmavinkurve-mdsol) [inline spec/lib/rack/content_type_default_spec.rb]: This example is very interesting and not intuitive. thanks for adding this one!
so we are overriding the `.json` route's content-type to be `xml` :(

i hope is not a real use-case. which is why i asked the question on the main PR about your use-case.

Comment (pmavinkurve-mdsol) [inline spec/lib/rack/content_type_default_spec.rb]: Jogging my memory here - i want to emphasize the difference between `Accept-header` and `Content-Type`.
For GETs content-type does not mean much as GETs are meant to have no body typically. 
so let's change this example here to a `post`. to make it more realistic. 
What do you think?

File: spec/lib/rack/content_type_default_spec.rb [hunk 17]
```diff
@@ -15,63 +17,117 @@
...
+      Rack::ContentTypeDefault.new(App, [:get, :post], 'application/xml', ['/authenticate.json']).call(post_req)
+      post_req['CONTENT_TYPE'].should == nil
+    end
+
+    it 'sets content type as appropriate on all requests if paths specified as all' do
+      Rack::ContentTypeDefault.new(App, :get, 'application/xml', 'all').call(get_req)
+      get_req['CONTENT_TYPE'].should == 'application/xml'
+    end
+
+    it 'sets content type as appropriate on all requests if paths is none are explicitly specified' do
```
Comment (pmavinkurve-mdsol) [inline spec/lib/rack/content_type_default_spec.rb]: minor: how about changing the second half `if paths is none are explicitly specified` ->
 `since no paths defaults to all paths`

File: spec/lib/rack/content_type_default_spec.rb [hunk 18]
```diff
@@ -15,63 +17,117 @@
...
+
+    it 'does not set content type on any requests if paths specified as empty ' do
+      get_req['PATH_INFO'] = 'api/v2/customers/2155/authenticate.json'
+      Rack::ContentTypeDefault.new(App, :get, 'application/xml', []).call(get_req)
+      get_req['CONTENT_TYPE'].should == nil
+    end
+  end
+
+  context 'default_based_on_path is set to true' do
+    it 'sets content type to application/json if path ends in .json' do
```
Comment (pmavinkurve-mdsol) [inline spec/lib/rack/content_type_default_spec.rb]: how about `overrides the content type` rather than `sets` to emphasize that even i specified `application/xml` i would get `.json`.

File: lib/rack/content_type_default.rb [hunk 19]
```diff
@@ -1,32 +1,54 @@
...
-
 module Rack
   class ContentTypeDefault
-    def initialize(app, methods = [:post], content_type = 'application/json')
+    def initialize(app, methods = [:post], content_type = 'application/json', paths = 'all', default_based_on_path = false)
       @app = app
       @methods = [*methods].map{ |item| item.to_s.upcase }
       @content_type = content_type
+      @paths = [*paths]
+      @default_based_on_path = default_based_on_path
```
Comment (pmavinkurve-mdsol) [inline lib/rack/content_type_default.rb]: how about `infer_based_on_path` for this variable rather than the word `default_`, or something shorter? can't think of anything as of now. basically we are giving it a hint. how's `use_path_hint`?

**Comment by pmavinkurve-mdsol** [comment]
@rcohen-mdsol finished reviewing. I think some comments ended up being blocking. 

Looking at your Jira story - primary_password filtering from logs - doesn't Rails provide a better way to do this??

File: lib/rack/content_type_default.rb (hunk 13 above)
Comment (pmavinkurve-mdsol) [inline lib/rack/content_type_default.rb]: it's funny - i was going to comment on the use of `or` in your original commit, but didn't. And today I came across the master RSG that says `and`, `or` keywords are banned :) 
https://github.com/bbatsov/ruby-style-guide#no-and-or-or 

Thanks for the updates @rcohen-mdsol !

File: README.md (hunk 10 above)
Comment (rcohen-mdsol) [inline README.md]: updated the README to describe the true/false option  @pmavinkurve-mdsol

**Comment by rcohen-mdsol** [comment]
Thanks for reviewing @pmavinkurve-mdsol Regarding your comments on the JIRA story, I've asked if there are available tools for this filtering issue on the DevTools forum. (It's important to note that I'm looking to filter from our logs before they are even pulled from by SumoLogic.) They mentioned an option for filtering from our logs (overriding the default filtering using mdsol_logging cookbook). However, that would not help in this specific situation because the sensitive information is already set to be filtered. 

(Also, the point about Rails doing the right magic for paths ending in .xml and .json was discussed on the JIRA story.)

**Comment by pmavinkurve-mdsol** [comment]
@rcohen-mdsol great! Sorry I am in a rush now, and getting ready to catch a long flight, but I looked at the updates real quick, and look pretty good. 

I am happy for @CGDS or @fpcyan or anyone else to go ahead and merge it if they are satisfied.
Thanks!

**Comment by rcohen-mdsol** [comment]
thank you @pmavinkurve-mdsol please review and merge if no further comments @CGDS @njacobs1 @fpcyan @cwhite-mdsol @sakhtermdsol

**Comment by cwhite-mdsol** [comment]
LGTM
